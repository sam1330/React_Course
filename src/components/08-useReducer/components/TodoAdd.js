import React from "react";
import PropTypes from 'prop-types';

import { useForm } from "../../../hooks/useForm";


const TodoAdd = ({ handleAddTodo }) => {

  const [{ description }, handleInputChange, reset] = useForm ({
    description: "",
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim() <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4 className="font-weight-bold">Add Todo</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          className="form-control"
          name="description"
          placeholder="Learn..."
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Add todo
        </button>
      </form>
    </>
  );
};

TodoAdd.propTypes = {
  handleAddTodo: PropTypes.func.isRequired
}

export default TodoAdd;