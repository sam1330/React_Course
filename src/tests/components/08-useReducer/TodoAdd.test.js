import { shallow } from "enzyme";
import React from "react";
import TodoAdd from "../../../components/08-useReducer/components/TodoAdd";

describe("<TodoAdd /> tests", () => {

  const handleAddTodo = jest.fn();

  const wrapper = shallow(
    <TodoAdd
      handleAddTodo={handleAddTodo}
    />
  )
  test('should render the component', () => {

    expect(wrapper).toMatchSnapshot();

  });

  test('should NOT call the handleAddTodo', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');

    formSubmit({ preventDefault() { } });

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('should call handleAddTodo', () => {

    const value = 'Learn React';

    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description'
      }
    });

    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() { } });

    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false
    });

    expect(wrapper.find('input').prop('value')).toBe('');

  });

});