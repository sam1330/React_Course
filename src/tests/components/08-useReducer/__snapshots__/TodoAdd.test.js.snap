// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoAdd /> tests should render the component 1`] = `
<Fragment>
  <h4
    className="font-weight-bold"
  >
    Add Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Todo..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
    >
      Add todo
    </button>
  </form>
</Fragment>
`;
